ha.md: explain how to prevent automatic reboots on HA-protected VMs (#459)
xcp-ng/xcp-ng-org: docs/management/ha.md

Reviewer: Not changed in this PR, but I find the phrasing ambiguous: "will restart the exact same way" leaves me unsure about the actual behaviour after a reboot.

However, if you halt the VM directly in the guest OS (via the console or in SSH), XCP-ng is NOT aware of what's going on. The system will think the VM is down and will consider that an anomaly. As a result, the VM will be **started automatically!**. This behavior prevents an operator from shutting down the system and leaving the VM unavailable for a long time.

#### Configure VM shutdown behavior

:::tip

Starting with XAPI 25.16.0, VM restart behavior can be changed on a pool-wide basis. To do this, run this command:

```
xe pool-param-set uuid=... ha-reboot-vm-on-internal-shutdown=false
```

> The `ha-reboot-vm-on-internal-shutdown` parameter indicates whether an HA-protected VM that is shut down from inside (not through the API) should be automatically rebooted when HA is enabled.

Reviewer: I have doubts about this "not through the API". 

There are two ways to shut a VM down via the API:
- soft shutdown (default) => transmits the shutdown order to the OS itself, which shuts down by itself
- hard shutdown => virtual electrical poweroff

How does HA behaves with regards to both these cases?

Author: I don't know. @psafont , do you know more about this?

Reviewer 2: The parameter is about shutdowns that are triggered from the guest OS (VM), in oposition to a xapi-initiated shutdown, I would word it differently:

> The `ha-reboot-vm-on-internal-shutdown` parameter indicates whether VM-initiated shudotwns will trigger a restart for HA-protected VMs, for example, when a user clicks the shutdown in Windows.

Reviewer: @psafont Do you know what happens with regards to HA if an admin triggers the VM shutdown from XAPI, either "soft" or "hard"?

Reviewer 2: From the wording in https://docs.xenserver.com/en-us/xenserver/8/high-availability, and from the inspection of code I've seen, the restarts are only done when the VM shuts down and xapi did not give the order.

This also explain this new field: it's a way to differentiate when a VM restarts from within, and when a VM has stopped because its host has become offline.

> The `ha-reboot-vm-on-internal-shutdown` parameter indicates whether VM-initiated shudotwns will trigger a restart for HA-protected VMs, for example, when a user clicks the shutdown in Windows.

Reviewer: ```suggestion
The `ha-reboot-vm-on-internal-shutdown` parameter indicates whether VM-initiated shutdowns will trigger a restart for HA-protected VMs, for example, when a user clicks the shutdown in Windows.
```

:::

:::warning

> Setting the `ha-reboot-vm-on-internal-shutdown` to `false` means that your VMs will stay off after they have been shut down from the guest OS, until you restart them yourself.

Reviewer: ```suggestion
Setting the `ha-reboot-vm-on-internal-shutdown` parameter to `false` means that your VMs will stay off after they have been shut down from the guest OS, until you restart them yourself.
```

> If you want to restore the default behavior (i.e. HA-protected VMs restart automatically after getting shut down from the guest OS), you will have to set the parameter to `true` again.

Reviewer: ```suggestion
If you want to restore the default behavior (i.e. HA-protected VMs restart automatically after getting shut down from the guest OS), set the parameter to `true` again.
```

:::

:::note

> If you don't want a specific VM to reboot automatically, you can also disable HA protection for that VM entirely. To do so, read the instructions at the [Troubleshooting HA section](../troubleshooting/troubleshooting-ha.md#disabling-ha).

Reviewer: ```suggestion
If you don't want a specific VM to reboot automatically, without changing the behavior for the whole pool, you can also temporarily disable HA protection for that VM. To do so, read the instructions at the [Troubleshooting HA section](../troubleshooting/troubleshooting-ha.md#disabling-ha).
```

:::

#### Configure VM shutdown behavior

If you don't want a VM to reboot automatically when it has been shut down from the guest OS:
1. Disable its HA protection.

> 2. Adjust the VM reboot behavior with the parameter called `Pool.ha_reboot_vm_on_internal_shutdown` (see below).

Reviewer 2: It's not clear by reading this section whether both need to be applied or only one of them.

Author: Fixed

:::warning

> Applying these changes on your entire VM pool means that your VMs will stay off after they have been shut down, until you restart them yourself. Your VMs will no longer be protected from accidental shutdowns. 

Reviewer 2: Does this warning apply to both options? I thought the first one was per-vm, and regarding the second one, I think mentioning that this is a VM-started shutdown is pretty important detail to mention, otherwise it makes it look like even a crash on a VM will need manual intervention to boot it up again.

Author: Are the warnings less ambiguous now?

:::

#### Configure VM shutdown behavior

> If you don't want a VM to reboot automatically when it has been shut down from the guest OS: disable its HA protection, then adjust the VM reboot behavior with the parameter called `Pool.ha_reboot_vm_on_internal_shutdown` (see below).

Reviewer: It was my understanding that you didn't have to both disable its HA protection and change the pool behaviour.

##### Disabling HA on specific VMs

You can adjust the reboot behavior for a specific HA-protected VM:

- **Using the `xe` cli:** :
    - To disable HA features, run `xe vm-param-set uuid=<vm_uuid> ha-restart-priority=`.
    - To (re-)enable HA features, run `xe vm-param-set uuid=<vm_uuid> ha-restart-priority=restart` or `xe vm-param-set uuid=<vm_uuid> ha-restart-priority=best-effort`.
- **Using Xen Orchestra**:
    To change the VM reboot behavior from Xen Orchestra, check out the instructions in the [Xen Orchestra documentation](https://docs.xen-orchestra.com/manage_infrastructure#vm-high-availability-ha).

:::warning
Applying these changes means that your VMs will stay off after they have been shut down, until you restart them yourself.
:::

:::tip
Once HA features have been disabled on your VM, shut the VM down. Once you have started the VM again, feel free to enable HA again.
:::

> ##### Disabling HA on the whole pool

Reviewer: The pool-wide setting isn't about disabling HA for the whole pool. Of course, disabling HA would work too, but here the question is how to govern HA's behaviour when it's on.

You can prevent automatic reboots on your entire pool. To do this, use the `xe` CLI to run this command:

> `xe pool-param-set uuid=$UUID ha-reboot-vm-on-internal-shutdown=false`.

Reviewer: So, actually, we already wrote about this on line 203. The online search didn't find "ha-reboot-vm" and my local grep of the files was on "ha_reboot_vm" so I missed it.

:::

##### For specific VMs

> If you don't want a VM to reboot automatically when it has been shut down from the guest OS, adjust the VM reboot behavior with the parameter called `Pool.ha_reboot_vm_on_internal_shutdown`.

Reviewer: That's the pool-wide setting! There's no per VM setting, all you can do is to change the `ha-restart-priority`.

:::warning

> Applying these changes means that your VMs will stay off after they have been shut down, until you restart them yourself.

Reviewer: This relates to the pool wide setting.
